### PR TITLE
Bulk update settings

### PR DESCRIPTION
**Pull request by tanyaxp**
This replaces 
https://github.com/tanyaxp/edx-platform/pull/1
#681

### CODE REVIEW

**Comment by tanyaxp** [comment]
stvstnfrd Please review my code. Thanks!

File: cms/envs/devstack.py [hunk 1]
```diff
@@ -62,6 +62,7 @@
 
 # By default don't use a worker, execute tasks as if they were local functions
 CELERY_ALWAYS_EAGER = True
+FEATURES['ENABLE_SERVICE_STATUS'] = True
```
Comment (stvstnfrd) [inline cms/envs/devstack.py]: I think we can remove this?

What does it do?

File: lms/envs/devstack.py [hunk 2]
```diff
@@ -17,7 +17,8 @@
 SITE_NAME = 'localhost:8000'
 PLATFORM_NAME = ENV_TOKENS.get('PLATFORM_NAME', 'Devstack')
 # By default don't use a worker, execute tasks as if they were local functions
-CELERY_ALWAYS_EAGER = True
+CELERY_ALWAYS_EAGER = False
```
Comment (stvstnfrd) [inline lms/envs/devstack.py]: This should presumably be switched back to `True`, right?

File: lms/envs/devstack.py [hunk 3]
```diff
@@ -17,7 +17,8 @@
...
+FEATURES['ENABLE_SERVICE_STATUS'] = True
```
Comment (stvstnfrd) [inline lms/envs/devstack.py]: Like the CMS equivalent, I think we can remove this?

File: openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+        /**
+         * Gets setting values from HTML
+         * @return {dict} Setting values the user wants changed, empty string
+         *                value if they don't want to change that setting
+         */
+        function getSettingsData() {
+            var data = {};
+            if ($applyMaxAttempts.is(':checked')) {
+                data.maxAttempts = parseInt($maxAttempts.val(),10);
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js]: Style: space after the comma.

File: openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                maxAttempts = data.maxAttempts,
+                showAnswer = data.showAnswer;
+
+            if (!maxAttempts && maxAttempts !== 0 && !showAnswer) {
+                return false;
+            }
+            if (maxAttempts && maxAttempts < 0) {
+                errors.push(VALIDATION_ERROR_MESSAGES.maxAttempts);
+            }
+            if (showAnswer && SHOW_ANSWER_OPTIONS.indexOf(showAnswer) < 0) {
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js]: Can we drop the left-hand side here?
This one does seem redundant..

File: openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+         * Displays error message on request error
+         */
+        function onError(response) {
+            var errMsg;
+            var serverMsg;
+            try {
+                serverMsg = $.parseJSON(response.responseText) || {};
+                errMsg = serverMsg.ErrMsg;
+            } catch (e) {
+                errMsg = '';
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js]: Should we display the exception text as the error message?

It seems a little weird that our error-handling in our error-handling doesn't display an error :)

File: openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            clearFeedbackDisplay();
+            if (validateData(data)) {
+                $submitBtn.prop('disabled', true);
+                $.ajax({
+                    type: 'POST',
+                    data: data,
+                    url: bulkupdateUrl,
+                    complete: onComplete,
+                    error: onError,
+                    success: onSuccess,
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/static/js/factories/bulkupdate.js]: Cool, this is looking good now.

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            '<input class="input setting-input setting-input-number" type="number" value="0" min="0.0000" step="1">'
+        )
+        self.assertContains(response, '<option value="always" selected>always</option>')
+        for option in SHOW_ANSWER_OPTIONS[1:]:
+            self.assertContains(response, '<option value="{}">{}</option>'.format(option, option))
+
+    @ddt.data('put','delete')
+    def test_bulkupdate_unsupported(self,method_name):
+        """
+        Operations is not supported (PUT, DELETE).
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: Nitpick: grammar?

`Operations is ...`

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        """
+        Tests email is sent to user after task completion
+        """
+        for count in range(len(SHOW_ANSWER_OPTIONS)):
+            settings = {'maxAttempts': count, 'showAnswer': SHOW_ANSWER_OPTIONS[count]}
+            response = self.client.post(self.bulkupdate_url, settings, HTTP_ACCEPT='application/json')
+            self.assertEqual(response.status_code, 200)
+            self.assertEqual(len(mail.outbox), count + 1)
+            self.assertEqual(
+                "{}'s bulk update on {}".format(self.user.username, self.course.display_name),
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: I think this is still testing the old way we were sending emails?

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/tasks.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+def _send_email_on_completion(course, user_fields, modified_settings, success):
+    """
+    Send email to user on completion of task, either success or failure
+    """
+    task_name = 'bulk settings update'
+    task_state_text = 'success'
+    dest_addr = user_fields['email']
+    try:
+        detail_url = reverse('utility_bulksettings_handler',kwargs={'course_key_string': course.id,})
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/tasks.py]: If we're going to use the `reverse_course_url` (or whatever) helper, we should use it everywhere.

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
Just a few minor things.
I didn't have a chance to run the tests locally, so please ask one of the other developers to try it out.

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
Just a few cleanup items

Remember to run quality checks for linting errors.

There also appears to be a bit of unused code following previous refactoring (unused imports, unused exception handling, etc.). I think I caught them all, but can you double-check?

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
+"""
+Unit tests for bulk update problem settings utility
+"""
+
+from django.core import mail
```
[the remark this answers is not in the document]
Comment (tanyaxp) [reply openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: Mail here is used to check if the test email has reached the outbox

**Pull request merged**